lib/index.js

```diff
@@ -152,37 +152,77 @@ export function configureUnifiedLanguageServer(
    * @param {boolean} alwaysStringify
    * @returns {Promise<VFile[]>}
    */
-  function processDocuments(textDocuments, alwaysStringify = false) {
-    return new Promise((resolve, reject) => {
-      engine(
-        {
-          alwaysStringify,
-          files: textDocuments.map((document) => lspDocumentToVfile(document)),
-          ignoreName,
-          packageField,
-          pluginPrefix,
-          plugins,
-          processor,
-          quiet: false,
-          rcName,
-          silentlyIgnore: true,
-          streamError: new PassThrough(),
-          streamOut: new PassThrough()
-        },
-        (error, _, context) => {
-          // An error never occur and can’t be reproduced. Thus us ab internal
-          // error in unified-engine. If a plugin throws, it’s reported as a
-          // vfile message.
-          /* c8 ignore start */
-          if (error) {
-            reject(error)
-          } else {
-            resolve((context && context.files) || [])
-          }
-          /* c8 ignore stop */
-        }
+  async function processDocuments(textDocuments, alwaysStringify = false) {
+    // LSP uses `file:` URLs (hrefs), `unified-engine` expects a paths.
+    // `process.cwd()` does not add a final slash, but `file:` URLs often do.
+    const workspacesAsPaths = [...workspaces].map((d) =>
+      fileURLToPath(d.replace(/\/$/, ''))
+    )
+    /** @type {Map<string, Array<VFile>>} */
+    const workspacePathToFiles = new Map()
+
+    if (workspacesAsPaths.length === 0) {
+      workspacesAsPaths.push(process.cwd())
+    }
+
+    for (const textDocument of textDocuments) {
+      const file = lspDocumentToVfile(textDocument)
+      const [cwd] = workspacesAsPaths
+        // Every workspace that includes the document.
+        .filter((d) => file.path.slice(0, d.length + 1) === d + path.sep)
+        // Sort the longest (closest to the file) first.
+        .sort((a, b) => b.length - a.length)
+
+      // This presumably should not occur: a file outside a workspace.
+      // So ignore the file.
+      /* c8 ignore next */
+      if (!cwd) continue
+
+      const files = workspacePathToFiles.get(cwd) || []
+      workspacePathToFiles.set(cwd, [...files, file])
+    }
+
+    /** @type {Array<Promise<Array<VFile>>>} */
+    const promises = []
+
+    for (const [cwd, files] of workspacePathToFiles) {
+      promises.push(
+        new Promise((resolve, reject) => {
+          engine(
+            {
+              alwaysStringify,
+              cwd,
+              files,
+              ignoreName,
+              packageField,
+              pluginPrefix,
+              plugins,
+              processor,
+              quiet: false,
+              rcName,
+              silentlyIgnore: true,
+              streamError: new PassThrough(),
+              streamOut: new PassThrough()
+            },
+            (error, _, context) => {
+              // An error never occur and can’t be reproduced. Thus us ab internal
```

Comment:
I know this is just an indentation change, but lets fix this typo while the line is touched.

```suggestion
              // An error never occur and can’t be reproduced. This is an internal
```

Reply:
Also `occur` -> `occur` 😅 I had a hard time reading this line lol.

I opened https://github.com/unifiedjs/unified-language-server/issues/24 to remove it all together soon!